api/src/test/java/marquez/FlowIntegrationTest.java

```diff
@@ -0,0 +1,185 @@
+import static org.assertj.core.api.Assertions.assertThat;
+
+import com.fasterxml.jackson.core.type.TypeReference;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableSet;
+import java.io.IOException;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
+import marquez.client.Utils;
+import marquez.client.models.Dataset;
+import marquez.client.models.DbTableMeta;
+import marquez.client.models.Field;
+import marquez.client.models.Job;
+import marquez.client.models.JobMeta;
+import marquez.client.models.Run;
+import marquez.client.models.RunMeta;
+import marquez.client.models.RunState;
+import marquez.common.models.FieldType;
+import org.apache.http.HttpResponse;
+import org.apache.http.client.methods.HttpGet;
+import org.apache.http.impl.client.CloseableHttpClient;
+import org.apache.http.impl.client.HttpClientBuilder;
+import org.apache.http.util.EntityUtils;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+
+@org.junit.jupiter.api.Tag("IntegrationTests")
+public class FlowIntegrationTest extends BaseIntegrationTest {
+
+  private static final String DATASET_NAME = "test-same-io-dataset";
+  private final CloseableHttpClient http = HttpClientBuilder.create().build();
+
+  @BeforeEach
+  public void setup() {
+    createNamespace(NAMESPACE_NAME);
+    createSource(DB_TABLE_SOURCE_NAME);
+    createSource(STREAM_SOURCE_NAME);
+  }
+
+  @Test
+  public void testSameInputAndOutputDatasetCreatedViaDatasetApi() throws IOException {
+
+    Dataset dataset = createDataset(null);
+
+    createJob(null);
+
+    Run createdRun = client.createRun(NAMESPACE_NAME, JOB_NAME, RunMeta.builder().build());
+
+    Dataset outputDataset = createDataset(createdRun.getId());
+
+    client.markRunAs(createdRun.getId(), RunState.COMPLETED);
+
+    assertInputDatasetVersionDiffersFromOutput(getRunResponse(createdRun));
+  }
+
+  @Test
+  public void testSameInputAndOutputDatasetWithJobRunIdUpdated() throws IOException {
+
+    Dataset dataset = createDataset(null);
+
+    createJob(null);
+
+    Run createdRun = client.createRun(NAMESPACE_NAME, JOB_NAME, RunMeta.builder().build());
+    Run runningState = client.markRunAs(createdRun.getId(), RunState.RUNNING);
+
+    createJob(createdRun.getId());
+
+    Run completed = client.markRunAs(createdRun.getId(), RunState.COMPLETED);
+
+    assertInputDatasetVersionDiffersFromOutput(getRunResponse(createdRun));
+  }
+
+  @Test
+  public void testSameInputAndOutputDatasetWithJobRunIdUpdatedAtTheEnd() throws IOException {
+
+    Dataset dataset = createDataset(null);
+
+    createJob(null);
+
+    Run createdRun = client.createRun(NAMESPACE_NAME, JOB_NAME, RunMeta.builder().build());
+    Run runningState = client.markRunAs(createdRun.getId(), RunState.RUNNING);
+    Run completed = client.markRunAs(createdRun.getId(), RunState.COMPLETED);
+    createJob(createdRun.getId());
```

Comment:
Is this order legal? I thought you needed to `createJob` _then_ `markRunAs(COMPLETED)`

Reply:
wanted to test even illegal order, cause no one stop user of doing that